js/__internal/grids/grid_core/views/m_columns_view.ts

```diff
@@ -1008,121 +1024,101 @@ export class ColumnsView extends viewWithColumnStateMixin {
-  }: any) {
-    let $cols;
-    let width;
-    let minWidth;
-    let columnIndex;
+  setColumnWidths({ widths, optionNames }: any): void {
+    const $tableElement = this.getTableElement();
+
+    if (!$tableElement?.length || !widths) {
+      return;
+    }
+
+    const columns = this.getColumns();
     const columnAutoWidth = this.option('columnAutoWidth');
 
-    $tableElement = $tableElement || this.getTableElement();
-
-    if ($tableElement && $tableElement.length && widths) {
-      columnIndex = 0;
-      $cols = $tableElement.children('colgroup').children('col');
-      setWidth($cols, 'auto');
-      columns = columns || this.getColumns(null, $tableElement);
-
-      for (let i = 0; i < columns.length; i++) {
-        if (columnAutoWidth && !fixed) {
-          width = columns[i].width;
-
-          if (width && !columns[i].command) {
-            width = columns[i].visibleWidth || width;
-
-            width = getWidthStyle(width);
-            minWidth = getWidthStyle(columns[i].minWidth || width);
-            // @ts-expect-error
-            const $rows = $rows || $tableElement.children().children('.dx-row').not(`.${DETAIL_ROW_CLASS}`);
-            for (let rowIndex = 0; rowIndex < $rows.length; rowIndex++) {
-              const row = $rows[rowIndex];
-
-              let cell;
-              const visibleIndex = this.getVisibleColumnIndex(i, rowIndex);
-              if (row.classList.contains(GROUP_ROW_CLASS)) {
-                cell = row.querySelector(`td[aria-colindex='${visibleIndex + 1}']:not(.${GROUP_CELL_CLASS})`);
-              } else {
-                cell = row.cells[visibleIndex];
-              }
-              if (cell) {
-                setCellWidth(cell, columns[i], width);
-                cell.style.minWidth = minWidth;
-              }
-            }
+    const $cols = $tableElement.children('colgroup').children('col');
+    setWidth($cols, 'auto');
+
+    columns.forEach((column, columnIndex) => {
+      if (columnAutoWidth && column.width && !column.command) {
+        const width = getWidthStyle(column.visibleWidth || column.width);
+        const minWidth = getWidthStyle(column.minWidth || width);
+
+        const $rows = $tableElement.children().children('.dx-row').not(`.${DETAIL_ROW_CLASS}`);
+
+        for (let rowIndex = 0; rowIndex < $rows.length; rowIndex++) {
+          const $row = $rows.eq(rowIndex);
+
+          const visibleIndex = this.getVisibleColumnIndex(columnIndex, rowIndex);
+
+          const $cell = $row.hasClass(GROUP_ROW_CLASS)
+            ? $row.find(`td[aria-colindex='${visibleIndex + 1}']:not(.${GROUP_CELL_CLASS})`)
+            : $row.find('td').eq(visibleIndex);
+
+          const cell = $cell.get(0);
+
+          if (cell) {
+            setCellWidth(cell, column, width);
+            cell.style.minWidth = minWidth;
           }
         }
+      }
 
-        if (columns[i].colspan) {
-          columnIndex += columns[i].colspan;
-          continue;
-        }
```

Comment:
Only band and fixed columns can have colspan. Method getColumns returns columns of the lastRow (in case of band columns used), so band columns never get here. Only if band column does not have any child columns, but in this case its colspan is 0. Thus, this if statement make sense only for fixed table